k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +103,277 @@ class SingleNsNamer(
+        // we are dealing with a numbered port. we will thus also
+        // need the port mappings from the `Service` API response,
+        // so join its activity with the endpoints cache activity.
+        cache.join(numberedPortRemappings(nsName, serviceName, labelSelector))
+          .map { case ((endpoints, ports)) =>
+            endpoints.lookupNumberedPort(ports, portNumber)
           }
-          Activity.Ok(paths)
-        }
-        Activity(states)
-      }
+      case None =>
+        // otherwise, we are dealing with a named port, so we can
+        // just look up the service from the endpointsCache
+        cache.map { endpoints => endpoints.port(portName) }
     }
+    stabilize(unstable).map { mkNameTree(id, residual) }
   }
 }
 
-private object EndpointsNamer {
-  case class Endpoint(ip: InetAddress, nodeName: Option[String])
+object EndpointsNamer {
+  val DefaultBackoff: Stream[Duration] =
+    Backoff.exponentialJittered(10.milliseconds, 10.seconds)
 
-  case class Svc(endpoints: Set[Endpoint], ports: Map[String, Int])
+  protected type PortMap = Map[String, Int]
+  protected type NumberedPortMap = Map[Int, String]
 
-  private[this] def getEndpoints(subsets: Option[Seq[v1.EndpointSubset]]): Set[Endpoint] = {
-    val endpoints = mutable.Set.empty[Endpoint]
-    for {
-      subset <- subsets.getOrElse(Seq.empty)
-      addresses <- subset.addresses
-      addrs <- addresses
-    } {
-      endpoints += Endpoint(InetAddress.getByName(addrs.ip), addrs.nodeName)
-    }
-    endpoints.toSet
-  }
+  protected case class Endpoint(ip: InetAddress, nodeName: Option[String])
 
-  private[this] def getPorts(subsets: Option[Seq[v1.EndpointSubset]]): Map[String, Int] = {
-    val portSet = mutable.Map.empty[String, Int]
-    for {
-      subset <- subsets.getOrElse(Seq.empty)
-      ports <- subset.ports
-      port <- ports
-      name <- port.name
-    } {
-      val proto = port.protocol.map(_.toUpperCase).getOrElse("TCP")
-      if (proto == "TCP") {
-        portSet(name) = port.port
-      }
-    }
-    portSet.toMap
+  protected object Endpoint {
+    def apply(addr: v1.EndpointAddress): Endpoint =
+      Endpoint(InetAddress.getByName(addr.ip), addr.nodeName)
   }
 
-  case class SvcCache(name: String, init: Svc) {
-
-    private[this] val endpointsState = Var[Set[Endpoint]](init.endpoints)
-    private[this] val portsState = Var[Map[String, Int]](init.ports)
-
-    def port(portName: String): Var[Option[Var[Addr]]] = {
-      portsState.map { portMap =>
-        val portNumber = portMap.get(portName)
-        portNumber.map(port)
-      }
-    }
-
-    def port(portNumber: Int): Var[Addr] =
-      endpointsState.map { endpoints =>
-        val addrs: Set[Address] = endpoints.map { endpoint =>
-          val isa = new InetSocketAddress(endpoint.ip, portNumber)
-          Address.Inet(isa, endpoint.nodeName.map(Metadata.nodeName -> _).toMap)
+  private[EndpointsNamer] case class EndpointsCache(
```

Comment:
`ServiceEndpoints` is a nice solution to the problem of having a whole million different things in this file whose names are some variation of "service" or "endpoints" 😛